SEO-670 | SEO friendly category pages

rogatty opened the pull request:
https://wikia-inc.atlassian.net/browse/SEO-670

@Wikia/core-platform-team @mszabo-wikia

Commit:
SEO-670 | load data for CategoryPage3

Commit:
SEO-670 | enable CategoryPage3 extension

Commit:
SEO-670 | render category members

Commit:
SEO-670 | load assets for category pages

Commit:
SEO-670 | add pagination

Commit:
SEO-670 | fix the first page url

Commit:
Merge branch 'dev' of github.com:Wikia/app into SEO-670

Commit:
SEO-670 | display thumbnails

Commit:
SEO-670 | subcategories don't have thumbnails displayed

Commit:
SEO-670 | display icon next to subcategories

Commit:
SEO-670 | display total number of members

Commit:
SEO-669 | update styles

Commit:
SEO-670 | temporarily log all sql queries

rogatty commented on extensions/wikia/CategoryPage3/CategoryPage3Model.class.php:
```diff
@@ -0,0 +1,284 @@
+	private function getMembersFromDB( DatabaseMysqli $dbr ): ResultWrapper {
+		$extraConditions = $this->getExtraConditionsForQuery( $dbr );
+
+		return $dbr->select(
```
Logs show it takes ~10-20 ms under normal load.

rogatty commented on extensions/wikia/CategoryPage3/CategoryPage3Model.class.php:
```diff
@@ -0,0 +1,284 @@
+	}
+
+	private function findTotalNumberOfMembers( DatabaseMysqli $dbr ) {
+		$totalNumberOfMembers = $dbr->selectField(
```
Logs show it takes ~10 ms for a category with ~44k members and ~1ms for a category with ~3k members under normal load.

rogatty commented on extensions/wikia/CategoryPage3/CategoryPage3Model.class.php:
```diff
@@ -0,0 +1,284 @@
+			return;
+		}
+
+		$res = $dbr->select(
```
Logs show it takes ~200 ms for a category with ~44k members and ~2-4ms for a category with ~3k members under normal load.

rogatty commented on extensions/wikia/CategoryPage3/CategoryPage3Model.class.php:
```diff
@@ -0,0 +1,284 @@
+	private function findLastPage( DatabaseMysqli $dbr ) {
+		$lastPageMembersCount = $this->totalNumberOfMembers % static::MEMBERS_PER_PAGE_LIMIT;
+
+		$res = $dbr->select(
```
Logs show it takes ~200 ms for a category with ~44k members and ~10ms for a category with ~3k members under normal load.

mszabo-wikia replied:
200 is a lot

rogatty replied:
Yup, that's why it's cached:
https://github.com/Wikia/app/blob/8504a5e27ec7cb65c6824920fe703020dad2749b/extensions/wikia/CategoryPage3/CategoryPage3Model.class.php#L242-L248

mszabo-wikia replied:
👍

rogatty commented on includes/db/Database.php:
```diff
@@ -203,7 +203,7 @@ function getServerInfo();
 
 	// @const log 5% of queries (increased from 1% in SUS-2974)
-	const QUERY_SAMPLE_RATE = 0.05;
+	const QUERY_SAMPLE_RATE = 1;
```
Don't forget about it.

Grunny commented on extensions/wikia/CategoryPage3/CategoryPage3Hooks.class.php:
```diff
@@ -0,0 +1,18 @@
+	 * @return bool
+	 */
+	static public function onArticleFromTitle( &$title, &$article ): bool {
+		if ( !$title || $title->getNamespace() !== NS_CATEGORY ) {
```
`!$title->inNamespace( NS_CATEGORY )` would be preferred.

Grunny commented on extensions/wikia/CategoryPage3/CategoryPage3Member.class.php:
```diff
@@ -0,0 +1,45 @@
+	}
+
+	public function isSubcategory(): bool {
+		return $this->getTitle()->getNamespace() === NS_CATEGORY;
```
`$this->getTitle()->inNamespace( NS_CATEGORY )`

Commit:
SEO-670 | cache the last page key and the number of members

Commit:
SEO-670 | encode the first character before rendering

Commit:
Merge branch 'dev' of github.com:Wikia/app into SEO-670

# Conflicts:
#	extensions/wikia/AssetsManager/config.php
#	includes/wikia/Extensions.php

Commit:
SEO-670 | set column break manually somewhere in the middle

Grunny commented on extensions/wikia/CategoryPage3/CategoryPage3CacheHelper.class.php:
```diff
@@ -0,0 +1,30 @@
+	 * @param $title
+	 * @return bool|Object
+	 */
+	public static function getTouched( $title ) {
```
I'd add a type here and on `setTouched`, given it would fail on `getKey` if it wasn't an instance of `Title` anyway, so I assume we can be strict.

Grunny commented on extensions/wikia/CategoryPage3/CategoryPage3Hooks.class.php:
```diff
@@ -0,0 +1,38 @@
+	 * @param Article $article
+	 * @return bool
+	 */
+	static public function onArticleFromTitle( &$title, &$article ): bool {
```
No need for the `&` on these, objects are effectively passed by reference in PHP regardless.

mszabo-wikia replied:
I think it's needed in this case for `$article` at least because we want to overwrite its value entirely

Grunny replied:
Ah, yeah, you're right, I missed that we were doing that here. So, yeah, only unnecessary on `$title`.

Commit:
SEO-670 | update styles

Commit:
SEO-670 | don't render anything for empty categories

Commit:
SEO-670 | update styles for pagination

Commit:
SEO-670 | hide section header when the first char is a space

Commit:
SEO-670 | use inNamespace method

Commit:
SEO-670 | use strict typing

Commit:
SEO-670 | remove obsolete reference

Commit:
Merge branch 'dev' of github.com:Wikia/app into SEO-670

Commit:
SEO-670 | display trending pages

Commit:
SEO-670 | add margin above trending pages header

Commit:
SEO-670 | use srcset to scale trending pages images per breakpoint

Commit:
SEO-670 | update design-system to 15.2.1

Commit:
SEO-670 | add dummy buttons for category layout selector

Commit:
SEO-670 | use design system typography variables

Commit:
SEO-670 | update styles for category layout selector

Commit:
SEO-670 | update styles for category layout selector

Commit:
SEO-670 | set cookie when category layout selector is clicked

Commit:
SEO-670 | add category layout selector to the mediawiki category page

Grunny committed:
(SEO-674) Redirect to the first category page when old params are used

Redirect to the first category page when old pagination query params are used.

Grunny committed:
(SEO-673) Add simple lazy loading for category item icons

Commit:
Merge pull request #16187 from Wikia/SEO-674

(SEO-674) Redirect to the first category page when old params are used

Commit:
Merge pull request #16193 from Wikia/SEO-673

(SEO-673) Simple LazyLoad implementation for testing

Commit:
SEO-670 | make CategoryExhibition compatible with new layout selector

Commit:
SEO-670 | remove CategoryExhibition from WikiFeatures

Commit:
SEO-670 | remove old query params if new category layout is used

Commit:
SEO-670 | remove redundant check for redirect

Commit:
SEO-670 | add category layout selector to global preferences

Commit:
SEO-670 | add i18n support

Commit:
SEO-670 | replace hardcoded strings with messages

Commit:
SEO-670 | add a tooltip for category layout selector

Commit:
Merge branch 'dev' of github.com:Wikia/app into SEO-670

# Conflicts:
#	includes/wikia/Extensions.php

Commit:
SEO-670 | update tooltip message

Commit:
SEO-670 | design review fixes

Commit:
SEO-670 | update design-system to 15.3.0

Commit:
SEO-670 | ignore browser cache on reload

Commit:
SEO-670 | skip cache after changing category layout

Commit:
SEO-670 | allow users to change sorting in Category Exhibition

Commit:
SEO-670 | replace user-generated pagination with non-crawlable links

Commit:
SEO-670 | split monolithic template into smaller ones

Commit:
SEO-670 | message for `Trending pages`

Grunny commented on extensions/wikia/CategoryPage3/CategoryPage3Hooks.class.php:
```diff
@@ -123,13 +123,51 @@ public static function onGetPreferences( $user, &$defaultPreferences ) {
+		$host = wfNormalizeHost( $host );
+
+		// External-external links
+		if ( !endsWith( $host, $wgFandomBaseDomain ) && !endsWith( $host, $wgWikiaBaseDomain ) ) {
```
This would pass for domains like `examplefandom.com`, right? It should check if it ends in `".{$wgFandomBaseDomain}"` as we do elsewhere to be safe.

Grunny commented on extensions/wikia/CategoryPage3/CategoryPage3Hooks.class.php:
```diff
@@ -123,13 +123,51 @@ public static function onGetPreferences( $user, &$defaultPreferences ) {
+		$categoryNsText = $wgContLang->getNsText( NS_CATEGORY );
+
+		// Not a category
+		if ( strpos( $path, $categoryNsText ) === false ) {
```
Maybe include the colon here just in case "Category" as a string appears in another page title.

Grunny commented on extensions/wikia/CategoryPage3/CategoryPage3Hooks.class.php:
```diff
@@ -170,4 +208,11 @@ private static function getLayoutForKey( Title $title, Article $article, $key )
+
+	private static function getNonCrawlableLink( $url, $text, $attribs ) {
+		$attribs['href'] = '#';
+		$attribs['onclick'] = 'window.location.assign("' . rawurlencode( $url ) . '"); return false;';
```
We should wrap the encoded URL in `Xml::escapeJsString`, just in case.

Commit:
SEO-670 | revert query sample rate change

Commit:
SEO-670 | improve checks and escaping

Grunny commented on extensions/wikia/CategoryPage3/CategoryPage3Hooks.class.php:
```diff
@@ -130,7 +130,10 @@ public static function onLinkerMakeExternalLink( string $url, string $text, stri
-		if ( !endsWith( $host, $wgFandomBaseDomain ) && !endsWith( $host, $wgWikiaBaseDomain ) ) {
+		if (
+			!endsWith( $host, ".${$wgFandomBaseDomain}" ) &&
+			!endsWith( $host, ".${$wgWikiaBaseDomain}" )
```
One too many `$`.

Grunny commented on extensions/wikia/CategoryPage3/CategoryPage3Hooks.class.php:
```diff
@@ -143,7 +146,7 @@ public static function onLinkerMakeExternalLink( string $url, string $text, stri
 
 		// Not a category
-		if ( strpos( $path, $categoryNsText ) === false ) {
+		if ( strpos( $path, "${$categoryNsText}:" ) === false ) {
```
One too many `$`. Also, can `$path` be URL encoded here? If so, should we account for any case in which `:` might be `%3A` (maybe just by unencoding `$path` for the comparison)?

Grunny commented on extensions/wikia/CategoryPage3/CategoryPage3Hooks.class.php:
```diff
@@ -210,8 +213,9 @@ private static function getLayoutForKey( Title $title, Article $article, $key )
+		$urlEscaped = Xml::escapeJsString( rawurlencode( $url ) );
 		$attribs['href'] = '#';
-		$attribs['onclick'] = 'window.location.assign("' . rawurlencode( $url ) . '"); return false;';
+		$attribs['onclick'] = "window.location.assign('${$urlEscaped}'); return false;";
```
One too many `$`.

mszabo-wikia commented on extensions/wikia/CategoryPage3/CategoryPage3Hooks.class.php:
```diff
@@ -0,0 +1,222 @@
+			$title = $article->getRedirectTarget();
+		}
+
+		if ( is_null( $title ) || !$title->inNamespace( NS_CATEGORY ) ) {
```
this null check is redundant, since the method call above will cause an error if `$title` is null

rogatty replied:
I think if there is a redirect but it's invalid, then it could be `null`.

mszabo-wikia commented on extensions/wikia/CategoryPage3/CategoryPage3Hooks.class.php:
```diff
@@ -0,0 +1,222 @@
+		return true;
+	}
+
+	public static function onLinkerMakeExternalLink( string $url, string $text, string &$link, array $attribs ): bool {
```
seems like there's some logic in here, do you think it'd be worth to add a unit test?

Commit:
SEO-670 | remove redundant $$$

Commit:
SEO-670 | fix double encoded url

Commit:
SEO-670 | add alphabet shortcuts

Commit:
SEO-670 | don't use columns when there is only one member

Commit:
SEO-670 | fix pagination styles on the last page

Commit:
SEO-670 | show the last page when `from` is larger than the last item

Commit:
SEO-670 | fix spacing between groups

Commit:
SEO-670 | use mixins for prefixable css properties

Commit:
SEO-670 | remove manual column break as it breaks layout on Edge

Commit:
SEO-670 | highlight the currently used alphabet shortcut

Commit:
SEO-670 | handle middle-click and ctrl+click on alphabet shortcuts

Commit:
Merge branch 'dev' of github.com:Wikia/app into SEO-670

Commit:
SEO-670 | improve handling of user generated alphabet shortcuts

Commit:
SEO-670 | expose new category pages in MercuryApi

Commit:
SEO-670 | :abc:

Commit:
SEO-670 | skip trending articles without thumbnail

Commit:
SEO-670 | pass total number of members in API

Commit:
SEO-670 | increase number of trending articles

Commit:
SEO-670 | fix wrong variable

Commit:
SEO-670 | encode alphabet shortcuts URLs for better crawling prevention

Commit:
SEO-670 | include all namespaces in category's trending pages

Commit:
SEO-670 | break long titles

Commit:
SEO-670 | include all namespaces in trending pages

Commit:
SEO-670 | include only content namespaces in trending pages

Commit:
SEO-670 | remove unused variable

Commit:
SEO-670 | VE works without any updates

Commit:
SEO-670 | check redirect only once

Commit:
SEO-670 | remove obsolete comment

$path is never encoded and if users bypass it by putting a full,
encoded URL in the link, then the link is broken by design

Commit:
SEO-670 | use surrogate keys in MercuryApi category responses

Commit:
SEO-670 | use mobile_wiki service to purge icache too

Commit:
SEO-670 | use mercury service as it's pre-configured

Commit:
SEO-670 | fix formatting

Commit:
SEO-670 | add rel=nofollow to uncrawlable links

Commit:
SEO-670 | download translations

Commit:
SEO-670 | remove obsolete comment

Commit:
SEO-670 | purge CDN too

Commit:
SEO-670 | load only the assets we need

Grunny commented on extensions/wikia/CategoryExhibition/CategoryExhibitionPage.php:
```diff
@@ -3,7 +3,31 @@
+	public function openShowCategory() {
+		parent::openShowCategory();
+
+		global $wgOut, $wgExtensionsPath, $wgJsMimeType, $wgTitle, $wgRequest, $wgUser;
```
```
$context = $this->getContext();
$context->getOutput();
$context->getTitle();
$context->getRequest();
$context->getUser();
```
^ would be preferred over the globals.

Grunny commented on extensions/wikia/CategoryPage3/CategoryPage3.class.php:
```diff
@@ -0,0 +1,194 @@
+				->width( $width )
+				->height( $height )
+				->url();
+		} catch ( Exception $e ) {
```
It looks like it only explicitly throws `InvalidArgumentException`, so being more specific here might be good to avoid potentially hiding other exceptions that we would want to know happened.

rogatty replied:
`->url()` throws `\Exception`:
https://github.com/Wikia/app/blob/6c60bc14c47ed41d2d72683c137e3c1f9acbaf27/lib/Wikia/src/Vignette/UrlGenerator.php#L283-L285

Grunny commented on extensions/wikia/CategoryPage3/CategoryPage3Model.class.php:
```diff
@@ -0,0 +1,320 @@
+				}
+			)
+		);
+		$imageServing = new ImageServing( $pageIds, $thumbWidth, array( 'w' => $thumbWidth, 'h' => $thumbHeight ) );
```
Minor: should use short array syntax per coding conventions.

Grunny commented on extensions/wikia/CategoryPage3/CategoryPage3TrendingPages.class.php:
```diff
@@ -0,0 +1,82 @@
+				if ( !empty( $item[ $param ] ) ) {
+					$processedItem[ $param ] = $item[ $param ];
+				} else {
+					return null;
```
Minor: is there any benefit to returning null rather than an empty array here? Would allow us to be strict about the return type rather than it being mixed/nullable, and the `empty` checks will still work.

Grunny commented on extensions/wikia/CategoryPage3/styles/category-layout-selector.scss:
```diff
@@ -0,0 +1,39 @@
+			}
+
+			&:last-child {
+				border-right: none;
```
Minor: I believe our [conventions](https://github.com/Wikia/guidelines/blob/47d00a7/SCSS/CodingConventions.md#borderzero) prefer `border-right: 0;` here.

Grunny commented on extensions/wikia/CategoryPage3/styles/category-page3.scss:
```diff
@@ -0,0 +1,195 @@
+
+			border-bottom: 1px solid $color-page-border;
+			display: grid;
+			grid-gap: 20px;
```
Does `grid-gap` work OK in Safari: https://developer.mozilla.org/en-US/docs/Web/CSS/gap#Browser_compatibility ?

rogatty replied:
From what I saw on @themech's Mac, it's fine. The standard property name is `gap` but all browsers I checked support `grid-gap` and Safari supports only `grid-gap`. That's why I used it here.

Grunny commented on extensions/wikia/CategoryPage3/styles/category-page3.scss:
```diff
@@ -0,0 +1,195 @@
+		&__alphabet-shortcuts {
+			font-weight: $wds-font-weight-bold;
+			list-style: none;
+			margin: 0 0 35px 0;
```
Could be `0 0 35px` I think.

Grunny commented on extensions/wikia/CategoryPage3/templates/members-header.php:
```diff
@@ -0,0 +1,15 @@
+<ul class="category-page__alphabet-shortcuts">
+	<?php foreach ( $alphabetShortcuts as $shortcut ) : ?>
+		<li class="category-page__alphabet-shortcut<?= $shortcut['isActive'] ? ' is-active': '' ?>">
+			<?= Html::rawElement( 'a', [
```
Can we use  `element` rather than `rawElement` to make sure it's escaped at the output?

Grunny commented on extensions/wikia/CategoryPage3/templates/pagination.php:
```diff
@@ -0,0 +1,27 @@
+		<?php endif; ?>
+		<?php if ( !empty( $pagination->getLastPageUrl() ) ): ?>
+            <a href="<?= $pagination->getLastPageUrl() ?>" class="wds-button wds-is-text">
+				<?= wfMessage( 'category-page3-pagination-last' )->escaped() ?>
```
Minor: whitespace seems very inconsistent in this file.

Grunny commented on extensions/wikia/CategoryPage3/templates/trending-pages.php:
```diff
@@ -0,0 +1,18 @@
+<div class="category-page__trending-pages-header"><?= wfMessage( 'category-page3-trending-pages' ) ?></div>
```
The default output format of `wfMessage` is `parse`. Since this message does not contain wikitext, to save using the parser, I'd explicitly use `->escaped()` here.

Commit:
SEO-670 | rename method

themech commented on extensions/wikia/CategoryPage3/CategoryPage3.class.php:
```diff
@@ -0,0 +1,194 @@
+			return;
+		}
+
+		$this->model->loadImages( 40, 30 );
```
please move these number to some consts. It will be easier to read the code, as `40, 30` doesn't tell me anything

rogatty replied:
![screenshot from 2018-11-02 13-27-12](https://user-images.githubusercontent.com/7030884/47915432-236a3a80-dea3-11e8-9c80-c79827f2021f.png)
Isn't that redundant? ;)

themech commented on extensions/wikia/CategoryExhibition/templates/form.tmpl.php:
```diff
@@ -1,40 +1,25 @@
-			$dropdown[] = $el;
+	<?= wfMessage( 'category-exhibition-sorttype' )->escaped(); ?>
+	<?php
+	$dropdown = array();
```
just a note: some of the code in this PR uses the `[]` syntax while there are still some places that use `array()`

themech commented on extensions/wikia/CategoryPage3/CategoryPage3CacheHelper.class.php:
```diff
@@ -0,0 +1,34 @@
+<?php
+
+class CategoryPage3CacheHelper {
+	const EXPIRATION_TIME = 60 * 60;
```
maybe something like `EXPIRATION_TIME_SEC`?

themech commented on extensions/wikia/CategoryPage3/CategoryPage3CacheHelper.class.php:
```diff
@@ -0,0 +1,34 @@
+		global $wgMemc;
+		$wgMemc->set(
+			static::getKey( $title ),
+			time() . rand( 0, 9999 ),
```
:)
I know, you just copied it from `CategoryExhibitionCacheHelper`

themech commented on extensions/wikia/CategoryPage3/CategoryPage3Hooks.class.php:
```diff
@@ -0,0 +1,221 @@
+	 * @return bool
+	 * @throws MWException
+	 */
+	public static function onAfterCategoriesUpdate( $categoryInserts, $categoryDeletes/*, $title*/ ): bool {
```
do we need this commented-out param?

rogatty replied:
I think that's a convention (one of many we have), in case someone wanted to do something with this param in the future.

themech commented on extensions/wikia/CategoryPage3/CategoryPageWithLayoutSelector.class.php:
```diff
@@ -0,0 +1,53 @@
+	 * @throws Exception
+	 */
+	protected function getCurrentLayout() {
+		throw new Exception( 'getCurrentLayout method needs to be overriden' );
```
So basically you want to define `CategoryPageWithLayoutSelector` as an abstract class and `getCurrentLayout` as an abstract method that has to be overriden, right?

themech commented on extensions/wikia/DesignSystem/node_modules/design-system/dist/css/styles.css:
```diff
@@ -1 +1 @@
-.wds-midlight-aqua,.wds-midlight-navy{-webkit-box-decoration-break:clone;box-decoration-break:clone;display:inline;padding-left:3px;padding-right:3px}.wds-avatar a,.wds-avatar__badge svg,.wds-avatar__image{height:100%;width:100%}.wds-font-weight-light{font-weight:300}.wds-font-weight-normal,.wds-global-footer__link,.wds-global-footer__link:active,.wds-global-footer__link:focus,.wds-global-footer_ ...
```
newline

rogatty replied:
This file is installed with npm, I'm not touching it.

themech commented on extensions/wikia/MercuryApi/models/MercuryApiCategoryModel.class.php:
```diff
@@ -1,5 +1,10 @@
+/**
+ * TODO Remove after SEO-670 is released on mobile-wiki and no cached JS tries to use it anymore
+ * @deprecated
+ * Class MercuryApiCategoryModel
```
👍 +1

themech commented on extensions/wikia/MercuryApi/handlers/MercuryApiMainPageHandler.class.php:
```diff
@@ -7,7 +7,11 @@ class MercuryApiMainPageHandler {
+		$trendingArticles = $mercuryApiModel->getTrendingPagesData(
+			10,
+			null,
+			false
```
it's not really obvious what these three values mean

rogatty replied:
![screenshot from 2018-11-02 13-38-44](https://user-images.githubusercontent.com/7030884/47915931-b8216800-dea4-11e8-8187-5067c99049ee.png)

![screenshot from 2018-11-02 13-39-38](https://user-images.githubusercontent.com/7030884/47915960-cc656500-dea4-11e8-99b7-158e49668406.png)

themech commented on extensions/wikia/CategoryPage3/CategoryPage3.class.php:
```diff
@@ -0,0 +1,194 @@
+			return;
+		}
+
+		$this->model->loadImages( 40, 30 );
```
same `loadImages( 40, 30 )` is used in another place, that makes me even more convinced that these two values should be moved to consts

Commit:
SEO-670 | add test for onLinkerMakeExternalLink

Commit:
SEO-670 | small clean up in CategoryExhibition

Commit:
SEO-670 | minor CR fixes

Commit:
Merge dev branch

Commit:
SEO-670 | add comment about CategoryTree dependency

Commit:
SEO-670 | use CategoryPage3 when CategoryExhibition isn't allowed

Commit:
SEO-670 | add tests for onArticleFromTitle

Commit:
SEO-670 | remove type from signature

Commit:
SEO-670 | use yield

Commit:
Merge dev branch

Commit:
SEO-670 | allow members per page to be overriden for testing purposes

Commit:
SEO-670 | fix for when the total number of items % items per page is 0

Commit:
SEO-670 | add integration test for category model

Commit:
SEO-670 | canonical should be consistent with pagination

Commit:
SEO-670 | local url is expected in MercuryApi

Commit:
SEO-670 | remove strict typing

Commit:
Merge branch 'dev' of github.com:Wikia/app into SEO-670
